### PR TITLE
[WIP] Add NucleusX Model

### PR DESCRIPTION
**Pull request by syncdoth**
# What does this PR do?

Add a new model named NucleusX Model.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/models/nucleus_x/test_modeling_nucleus_x.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        self.config_tester = ConfigTester(self, config_class=NucleusXConfig, decoder_embed_dim=64)
+
+    def test_config(self):
+        self.config_tester.run_common_tests()
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_model_various_embeddings(self):
```
Comment (syncdoth) [inline tests/models/nucleus_x/test_modeling_nucleus_x.py]: delete this test case.

File: src/transformers/models/nucleus_x/configuration_nucleus_x.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        bos_token_id (`int`, *optional*, defaults to 1):
+            Beginning of stream token id.
+        eos_token_id (`int`, *optional*, defaults to 2):
+            End of stream token id.
+        use_cache (`bool`, *optional*, defaults to `True`):
+            Whether or not the model should return the last key/values attentions (not used by all models). Only
+            relevant if `config.is_decoder=True`.
+        forward_mode (`str`, *optional*, defaults to `"parallel"`):
+            The three modes of Retentive Network: `"parallel"`, `"recurrent"`, or `"chunkwise"`.
+        activation_fn (`str` or `function`, *optional*, defaults to `"gelu"`):
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/configuration_nucleus_x.py]: We default to `swish` activation.

File: src/transformers/models/nucleus_x/configuration_nucleus_x.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    attribute_map = {
+        "hidden_size": "decoder_embed_dim",
+        "intermediate_size": "decoder_ffn_embed_dim",
+        "num_attention_heads": "decoder_retention_heads",
+        "num_hidden_layers": "decoder_layers",
+    }
+
+    def __init__(
+        self,
+        vocab_size: int = 32000,
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/configuration_nucleus_x.py:139]: Are we sure we pre-trained our models with `vocab_size = 32000`? I think we have a bug with `50257` in our pre-training. Can you check this?

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py [hunk 4]
```diff
@@ -0,0 +1,1410 @@
...
+            )
+            # decay_scale (used for kv cache)
+            if get_decay_scale:
+                decay_scale = self.compute_decay_scale(slen, retention_mask)
+            else:
+                decay_scale = None
+            retention_rel_pos = (
+                (sin, cos),
+                (inner_mask, cross_decay, query_inner_decay, value_inner_decay, decay_scale),
+            )
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/modeling_nucleus_x.py]: Nothing major here. Please move this block to a different function. Long functions are less readable.

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py [hunk 5]
```diff
@@ -0,0 +1,1410 @@
...
+        # layers * [bsz, num_head, qk_dim, decoder_embed_dim]
+        next_decoder_cache = () if use_cache else None
+
+        for idx, layer in enumerate(self.layers):
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            past_key_value = past_key_values[idx] if past_key_values is not None else None
+
+            if self.gradient_checkpointing and self.training:
+
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/modeling_nucleus_x.py:1065]: Please add a logic to make sure use_cache is `False`
Ref: https://github.com/huggingface/transformers/blob/5125fb9aa0803580995445561c1d5375f5f79c56/src/transformers/models/falcon/modeling_falcon.py#L776

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py [hunk 6]
```diff
@@ -0,0 +1,1410 @@
...
+        - decay_mask: [(1 or bsz) * num_head * len * len]
+        """
+        decay_mask, intra_decay, scale = decay_mask
+
+        # [b, h, t, t]
+        retention = q @ k.transpose(-1, -2)  # (scaled dot-product)
+        retention = retention * decay_mask
+
+        # invariant after normalization
+        retention = retention / retention.detach().sum(dim=-1, keepdim=True).abs().clamp(min=1)
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/modeling_nucleus_x.py:312]: Is `detach()` necessary here?? Is there a solid reason why this would make sense?

Comment (sippycoder) [reply src/transformers/models/nucleus_x/modeling_nucleus_x.py:312]: I see that this is used in the torch scale implementation as well. So weird.

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py [hunk 7]
```diff
@@ -0,0 +1,1410 @@
...
+        # [b, n_c, h, t_c, qkv_dim]
+        q = q.view(bsz, self.num_heads, num_chunks, chunk_len, self.key_dim).transpose(1, 2)
+        k = k.view(bsz, self.num_heads, num_chunks, chunk_len, self.key_dim).transpose(1, 2)
+        v = v.view(bsz, self.num_heads, num_chunks, chunk_len, self.head_dim).transpose(1, 2)
+
+        k_t = k.transpose(-1, -2)
+
+        qk_mat = q @ k_t  # [b, n_c, h, t_c, t_c]
+        qk_mat = qk_mat * decay_mask.unsqueeze(1)
+        inner_scale = qk_mat.detach().abs().sum(dim=-1, keepdim=True).clamp(min=1)
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/modeling_nucleus_x.py:401]: Why does using `detach()` make sense?

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py [hunk 8]
```diff
@@ -0,0 +1,1410 @@
...
+        cross_scale = []
+        kv_state = torch.zeros(bsz, self.num_heads, self.key_dim, self.head_dim).to(v)
+        kv_scale = torch.ones(bsz, self.num_heads, 1, 1).to(v)
+
+        # accumulate kv by loop
+        for i in range(num_chunks):
+            kv_recurrent.append(kv_state / kv_scale)
+            cross_scale.append(kv_scale)
+            kv_state = kv_state * cross_decay + kv[:, i]
+            kv_scale = kv_state.detach().abs().sum(dim=-2, keepdim=True).max(dim=-1, keepdim=True).values.clamp(min=1)
```
Comment (sippycoder) [inline src/transformers/models/nucleus_x/modeling_nucleus_x.py:421]: `detach()` is used here again.

**Review by sippycoder — COMMENTED** [review]
About 60% reviewed. Will continue.

File: src/transformers/models/nucleus_x/configuration_nucleus_x.py (hunk 3 above)
Comment (syncdoth) [reply src/transformers/models/nucleus_x/configuration_nucleus_x.py:139]: In our train code, we reset the vocab size based on the tokenizer we use. So our models will have vocab size of 32000 (or 32001 for 7b, including pad_token. For 13b, I'm using unk_token as pad_token following openllama.)

File: src/transformers/models/nucleus_x/modeling_nucleus_x.py (hunk 6 above)
Comment (syncdoth) [reply src/transformers/models/nucleus_x/modeling_nucleus_x.py:312]: I have no solid idea why we need detach. But I guess this here is for pure scaling purposes for numerical stability, so they don't extra want gradient operations here.
